tick/preprocessing/longitudinal_features_lagger.py

```diff
@@ -0,0 +1,250 @@
+            [1, n_intervals]. If the value i is equal to n_intervals, then there
+            is no censoring for sample i. If censoring = c < n_intervals, then 
+            the observation of sample i is stopped at interval c, that is, the 
+            row c - 1 of the correponding matrix. The last n_intervals - c rows
+            are then set to 0.
+
+        Returns
+        -------
+        output : `LongitudinalFeaturesProduct`
+            The fitted current instance.
+        """
+        self._reset()
+        base_shape = X[0].shape
+        X = LFP._check_X_consistency(X, base_shape, "float64")
+        n_intervals, n_init_features = base_shape
+        self._set("_n_init_features", n_init_features)
+        self._set("_n_intervals", n_intervals)
+        mapper = {i: tuple(i + j for j in range(self.n_lags + 1))
+                  for i in range(self._n_init_features)}
+        self._set("_mapper", mapper)
+        self._set("_n_output_features", int(n_init_features *
+                                            (self.n_lags + 1)))
+        self._set("_cpp_preprocessor",
+                  _LongitudinalFeaturesLagger(X, self.n_lags))
+        self._set("_fitted", True)
+
+        return self
+
+    def transform(self, X, censoring):
+        """Add lagged features to the given features matrices list.
+
+        Parameters
+        ----------
+        X : list of np.ndarray or list of scipy.sparse.csr_matrix,
+            list of length n_samples, each element of the list of 
+            shape=(n_intervals, n_features)
+            The list of features matrices.
+
+        censoring : `np.ndarray`, shape=(n_samples, 1), dtype="uint64"
+            The censoring data. This array should contain integers in 
+            [1, n_intervals]. If the value i is equal to n_intervals, then there
+            is no censoring for sample i. If censoring = c < n_intervals, then 
+            the observation of sample i is stopped at interval c, that is, the 
+            row c - 1 of the correponding matrix. The last n_intervals - c rows
+            are then set to 0.
+
+        Returns
+        -------
+        output : `[np.ndarrays]`  or `[csr_matrices]`, shape=(n_intervals, n_features)
+            The list of features matrices with added lagged features.
+        """
+
+        n_samples = len(X)
+        censoring = self._check_censoring_consistency(censoring, n_samples)
+        base_shape = (self._n_intervals, self._n_init_features)
+        X = LFP._check_X_consistency(X, base_shape, "float64")
+        if sps.issparse(X[0]):
+            X_with_lags = [self._sparse_lagger(x, int(censoring[i]))
+                           for i, x in enumerate(X)]  # Don't get why int() is required here as censoring_i is uint64
+        else:
+            X_with_lags = [self._dense_lagger(x, int(censoring[i]))
+                           for i, x in enumerate(X)]
+
+        return X_with_lags
+
```

Comment:
It adds lags, i.e. add new columns to each element of the list. These elements are ndarrays or csr_matrices, so it seems as inefficient to rebuild the objects than to extend them: ndarrays will trigger copies at some point to be contiguous in memory, and it cannot be done with scipy.sparse.csr_matrices